File: backend-go/internal/pipeline/cluster.go
package pipeline

import (
	"context"
	"time"

	"github.com/worldsignal/backend/internal/db"
	"github.com/worldsignal/backend/internal/textutil"
)

const (
	similarityThreshold = 0.5
	windowHours         = 72
)

// ClusterResult mirrors the TS ClusterResult.
type ClusterResult struct {
	SignalID string
	IsNew    bool
}

// ClusterArticle attaches an article to an existing Signal or creates a new one,
// using token-set Jaccard similarity over a recent window. Mirrors cluster.ts.
// `now` is injected for determinism in tests.
func ClusterArticle(ctx context.Context, d *db.DB, articleID string, now time.Time) (*ClusterResult, error) {
	article, err := d.GetClusterArticle(ctx, articleID)
	if err != nil || article == nil {
		return nil, err
	}

	existing, err := d.ExistingSignalForArticle(ctx, articleID)
	if err != nil {
		return nil, err
	}
	if existing != "" {
		return &ClusterResult{SignalID: existing, IsNew: false}, nil
	}

	since := now.Add(-windowHours * time.Hour)
	candidates, err := d.RecentSignalCandidates(ctx, since)
	if err != nil {
		return nil, err
	}

	bestID := ""
	bestScore := 0.0
	first := true
	for _, c := range candidates {
		score := textutil.Jaccard(article.TokenSet, c.TokenSet)
		if first || score > bestScore {
			bestID = c.ID
			bestScore = score
			first = false
		}
	}

	if bestID != "" && bestScore >= similarityThreshold {
		if err := d.AttachArticleToSignal(ctx, bestID, articleID, bestScore, now); err != nil {
			return nil, err
		}
		return &ClusterResult{SignalID: bestID, IsNew: false}, nil
	}

	id, err := d.CreateSignalFromArticle(ctx, article, now)
	if err != nil {
		return nil, err
	}
	return &ClusterResult{SignalID: id, IsNew: true}, nil
}

File: backend/src/pipeline/cluster.ts
import { prisma } from "../db/prisma.js";
import { jaccard } from "../lib/text.js";
import { logger } from "../lib/logger.js";

const log = logger("cluster");

const SIMILARITY_THRESHOLD = 0.5;
const WINDOW_HOURS = 72;

export interface ClusterResult {
  signalId: string;
  isNew: boolean;
}

/**
 * Attach an Article to an existing Signal (same event) or create a new one.
 * Lightweight token-set Jaccard similarity over a recent window — no vector DB,
 * Postgres only.
 */
export async function clusterArticle(articleId: string): Promise<ClusterResult | null> {
  const article = await prisma.article.findUnique({ where: { id: articleId } });
  if (!article) return null;

  // Already linked? (idempotency on retried jobs)
  const existingLink = await prisma.signalArticle.findFirst({ where: { articleId } });
  if (existingLink) return { signalId: existingLink.signalId, isNew: false };

  const since = new Date(Date.now() - WINDOW_HOURS * 3600 * 1000);
  const candidates = await prisma.signal.findMany({
    where: { lastSeenAt: { gte: since } },
    orderBy: { lastSeenAt: "desc" },
    take: 300,
    select: { id: true, metadata: true, sourceCount: true },
  });

  const articleTokens = article.tokenSet ?? "";
  let best: { id: string; score: number } | null = null;
  for (const c of candidates) {
    const meta = (c.metadata ?? {}) as { tokenSet?: string };
    const score = jaccard(articleTokens, meta.tokenSet ?? "");
    if (!best || score > best.score) best = { id: c.id, score };
  }

  const now = new Date();
  if (best && best.score >= SIMILARITY_THRESHOLD) {
    await prisma.$transaction([
      prisma.signalArticle.create({
        data: {
          signalId: best.id,
          articleId,
          relationType: "SUPPORTING",
          similarityScore: best.score,
        },
      }),
      prisma.signal.update({
        where: { id: best.id },
        data: {
          sourceCount: { increment: 1 },
          lastSeenAt: now,
        },
      }),
    ]);
    log.debug(`article ${articleId} joined signal ${best.id} (score ${best.score.toFixed(2)})`);
    return { signalId: best.id, isNew: false };
  }

  const seen = article.publishedAt ?? now;
  const signal = await prisma.signal.create({
    data: {
      title: article.title,
      summary: article.summary ?? article.title,
      status: "UNVERIFIED",
      firstSeenAt: seen,
      lastSeenAt: now,
      country: article.country,
      sourceCount: 1,
      metadata: { tokenSet: articleTokens },
      articles: {
        create: { articleId, relationType: "PRIMARY", similarityScore: 1 },
      },
    },
  });
  log.debug(`article ${articleId} created new signal ${signal.id}`);
  return { signalId: signal.id, isNew: true };
}
